### DISCUSSION TITLE
Enforcing Consistency of Related Entities in Prisma Data Model

### DISCUSSION THREAD

**Discussion by laurent512** [prisma/prisma#20748] [Q&A] [answered]
Hi everyone, 

I have a schema involving three models: Repartition, Coproperty, and Lot. 

```prisma
model Repartition {
  id            String     @id @default(cuid())
  description   String?
  coproperty    Coproperty @relation(fields: [coproperty_id], references: [id])
  coproperty_id String
  paying_lots   Lot[]
}

model Coproperty {
  id           String        @id @default(cuid())
  Repartition  Repartition[]
  lots         Lot[]
}

model Lot {
  id           String        @id @default(cuid())
  Coproperty   Coproperty    @relation(fields: [copropertyId], references: [id])
  copropertyId String
  members      LotMember[]
  stake        Int 
  repartitions Repartition[]
}
```

I want to ensure that all the paying_lots associated with a Repartition are related to the same coproperty (that is the one mentionned in the repartition) . How can I enforce this constraint either at the database schema level or through application logic using Prisma?

**Comment by ludralph** [comment, marked as answer]
Hi @laurent512 👋 

You can enforce the constraint that all the `paying_lots` associated with a `Repartition` must be related to the same `Coproperty` using both database schema design and application logic. In your schema, there is a `coproperty_id` field in the Repartition model that establishes a relationship with the `Coproperty` model. This association is what you can leverage to enforce the constraint in your database.
In your application logic, you can create a custom validation step to ensure that the `paying_lots` associated with a `Repartition` all belong to the same `Coproperty`. For example:
```js

async function createRepartitionWithLots(repartitionData, lotIds) {
  const { coproperty_id } = repartitionData;

  // Fetch the Coproperty associated with the provided coproperty_id
  const coproperty = await prisma.coproperty.findUnique({
    where: { id: coproperty_id },
    include: { lots: true }, // Include lots associated with the Coproperty
  });

  if (!coproperty) {
    throw new Error("Coproperty not found");
  }

  // Check if all provided lotIds belong to the same Coproperty
  const lotsBelongToSameCoproperty = lotIds.every(lotId =>
    coproperty.lots.some(lot => lot.id === lotId)
  );

  if (!lotsBelongToSameCoproperty) {
    throw new Error("Lots must belong to the same Coproperty");
  }

  // Proceed with creating the Repartition and associating paying_lots
  const repartition = await prisma.repartition.create({
    data: {
      ...repartitionData,
      paying_lots: {
        connect: lotIds.map(lotId => ({ id: lotId })),
      },
    },
  });

  return repartition;
}

// Usage example
const repartitionData = {
  coproperty_id: "your-coproperty-id",
  description: "Some description",
};

const lotIds = ["lot-id-1", "lot-id-2", "lot-id-3"]; // Lot IDs belonging to the same Coproperty

createRepartitionWithLots(repartitionData, lotIds)
  .then(repartition => {
    console.log("Repartition created:", repartition);
  })
  .catch(error => {
    console.error("Error:", error.message);
  });
```

In this example, the `createRepartitionWithLots` function performs a custom validation step to ensure that all the provided `lotIds` belong to the same `Coproperty` before creating the `Repartition` and associating the `paying_lots`. If the validation fails, an error is thrown. With the code above, you can enforce the constraint that all paying_lots associated with a `Repartition` must be related to the same `Coproperty`.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
